src/helm/proxy/clients/together_client.py

```diff
@@ -72,28 +87,84 @@ def __init__(self, cache_config: CacheConfig, api_key: Optional[str] = None):
-
-        response, cached = self.cache.get(cache_key, wrap_request_time(do_it if self.api_key else fail))
+        if not self.api_key:
+            raise TogetherClientError("togetherApiKey not set in credentials.conf")
+        headers: Dict[str, str] = {"Authorization": f"Bearer {self.api_key}"}
+
+        if request.model_engine in _ASYNC_MODELS:
+
+            def submit_job() -> str:
+                submit_request = raw_request.copy()
+                submit_request["async"] = True
+                submit_response = requests.post(TogetherClient.INFERENCE_ENDPOINT, headers=headers, json=submit_request)
+                try:
+                    submit_response.raise_for_status()
+                except Exception as e:
+                    raise TogetherClientError(
+                        f"Together job submission request failed with {submit_response.status_code}: "
+                        f"{submit_response.text}"
+                    ) from e
+                submit_response_json = submit_response.json()
+                job_id = submit_response_json.get("id")
+                if not job_id:
+                    raise TogetherClientError(
+                        f"Could not get job_id from job submission response {submit_response_json}"
+                    )
+                return job_id
+
+            # Retry with a 1 second delay that doubles every retry until a maximum of delay of 15 seconds.
+            # Stop retrying after 1 minute.
+            @retry(wait_exponential_multiplier=1000, wait_exponential_max=15000, stop_max_delay=60000)
+            def retrieve_job(job_id: str) -> Dict[Any, Any]:
+                job_url = self._get_job_url(job_id)
+                retrieve_response = requests.get(job_url, headers=headers)
+                try:
+                    retrieve_response.raise_for_status()
+                except Exception as e:
+                    raise TogetherClientError(
+                        f"Together job retrieval request failed with {retrieve_response.status_code}: "
+                        f"{retrieve_response.text}"
+                    ) from e
+                retrieve_response_json = retrieve_response.json()
+                if retrieve_response_json["status"] != "finished":
+                    raise TogetherClientError(
+                        f"Together job did not finish in {TogetherClient.RETRIEVE_JOB_MAX_WAIT_SECONDS} seconds: "
+                        f"{job_id}"
+                    )
+                if "output" not in retrieve_response_json:
+                    raise TogetherClientError(
+                        f"Could not get output from Together job {job_id}: {retrieve_response_json}"
+                    )
+                return retrieve_response_json["output"]
+
+            def do_it_async() -> Dict[Any, Any]:
+                job_id = submit_job()
+                return retrieve_job(job_id)
```

Comment:
This is cool / slick, but I wonder if it'd be clearer / more customizable if we just wrote a while loop and can better control the output...?